web/src/compose_validate.ts

```diff
@@ -329,6 +330,49 @@ export function warn_if_in_search_view(): void {
     }
 }
 
+export function clear_guest_in_dm_recipient_warning(): void {
+    const classname = compose_banner.CLASSNAMES.guest_in_direct_message_recipient;
+    $(`#compose_banners .${CSS.escape(classname)}`).remove();
+}
+
+// Only called on recipient change. Adds new banner if not already
+// exists or updates the existing banner or remove banner is no
+// guest in the dm.
+export function warn_if_guest_in_dm_recipient(): void {
+    const email_string = compose_state.private_message_recipient();
+    const guest_emails = people.filter_guest_emails(email_string);
+
+    // Only messages of type private are valid
+    const is_valid = compose_state.get_message_type() === "private";
+
+    if (!realm.realm_enable_guest_user_dm_warning || !is_valid || !guest_emails) {
+        clear_guest_in_dm_recipient_warning();
+        return;
+    }
+
+    const classname = compose_banner.CLASSNAMES.guest_in_direct_message_recipient;
+    let $banner = $(`#compose_banners .${CSS.escape(classname)}`);
+    const banner_text = people.get_dm_guest_banner_text(guest_emails);
+
+    if (compose_state.get_recipient_guest_emails() === guest_emails.join(",")) {
+        return;
+    }
+    compose_state.set_recipient_guest_emails(guest_emails.join(","));
+
+    if ($banner.length === 1) {
+        $banner.find(".banner_content").text(banner_text);
+        return;
+    }
+
+    $banner = $(
+        render_guest_in_dm_recipient_warning({
+            banner_text,
+            classname: compose_banner.CLASSNAMES.guest_in_direct_message_recipient,
+        }),
+    );
+    compose_banner.append_compose_banner_to_banner_list($banner, $("#compose_banners"));
+}
```

Comment:
Except the above comment, I think this function could use some more work to make it more easier to read.

Reply:
Added some more comments in the function.